[WebGPU] setPipelineState fails when X32_Stencil8 attachment is used in a Stencil8 pipeline

mwyrzykowski opened the pull request:
#### b2b40994e9d80dd379273b97f5e33c99fa8c0e0d
<pre>
[WebGPU] setPipelineState fails when X32_Stencil8 attachment is used in a Stencil8 pipeline
<a href="https://bugs.webkit.org/show_bug.cgi?id=273021">https://bugs.webkit.org/show_bug.cgi?id=273021</a>
&lt;radar://126621251&gt;

Reviewed by Tadeu Zagallo.

Calling setPipelineState with a Stencil8 format will fail so check
if we have such a mismatch before passing the validation check.

* LayoutTests/TestExpectations:
* LayoutTests/fast/webgpu/fuzz-273021-expected.txt: Added.
* LayoutTests/fast/webgpu/fuzz-273021.html: Added.
Add regression test.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::clearTextureIfNeeded):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::colorDepthStencilTargetsMatch const):

Canonical link: <a href="https://commits.webkit.org/277856@main">https://commits.webkit.org/277856@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/333846c5bc9bf228b1e4e7168127a214b6db6157

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48810 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28022 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51776 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51496 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44876 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33965 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25551 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/39899 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49392 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25670 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/42086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21000 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23131 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/43267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6865 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/45064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/43756 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53407 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23861 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/20131 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47201 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/25124 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42293 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46146 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10744 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25931 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/24844 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -727,7 +727,7 @@ static bool isRenderableTextureView(const TextureView& texture)
     ASSERT(!m_existingCommandEncoder);
     setExistingEncoder(mtlRenderCommandEncoder);
-    return RenderPassEncoder::create(mtlRenderCommandEncoder, descriptor, visibilityResultBufferSize, depthReadOnly, stencilReadOnly, *this, visibilityResultBuffer, maxDrawCount, m_device);
+    return RenderPassEncoder::create(mtlRenderCommandEncoder, descriptor, visibilityResultBufferSize, depthReadOnly, stencilReadOnly, *this, visibilityResultBuffer, maxDrawCount, mtlDescriptor.stencilAttachment.texture.pixelFormat == MTLPixelFormatX32_Stencil8, m_device);
```
I'm just curious, but would it make more sense to pass the pixelFormat instead of a boolean?

mwyrzykowski replied:
oh that's a good idea, but actually I realized we already have this information, I don't need to pass another parameter.

On RenderPassEncoder.cpp:135 we have:
```
id<MTLTexture> depthTexture = textureView.texture();
```
I can get the pixel format from there

mwyrzykowski replied:
PR is updated and much simpler now :)

mwyrzykowski committed:
[WebGPU] setPipelineState fails when X32_Stencil8 attachment is used in a Stencil8 pipeline
https://bugs.webkit.org/show_bug.cgi?id=273021
<radar://126621251>

Reviewed by Tadeu Zagallo.

Calling setPipelineState with a Stencil8 format will fail so check
if we have such a mismatch before passing the validation check.

* LayoutTests/TestExpectations:
* LayoutTests/fast/webgpu/fuzz-273021-expected.txt: Added.
* LayoutTests/fast/webgpu/fuzz-273021.html: Added.
Add regression test.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::clearTextureIfNeeded):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::colorDepthStencilTargetsMatch const):

Canonical link: https://commits.webkit.org/277856@main
